src/ir/display.rs

```diff
@@ -3,18 +3,47 @@
 use super::{FuncDecl, FunctionBody, Module, SourceLoc, ValueDef};
 use crate::entity::EntityRef;
 use std::collections::HashMap;
-use std::fmt::{Display, Formatter, Result as FmtResult};
+use std::fmt::{self, Display, Formatter, Result as FmtResult};
+
+/// Hooks to print information after inst, before and after blocks
+/// and before and after functions.
+pub trait PrintDecorator {
+    fn after_inst(&self, _value: super::Value, _f: &mut fmt::Formatter) -> fmt::Result {
+        Ok(())
+    }
+
+    fn before_block(&self, _block: super::Block, _f: &mut fmt::Formatter) -> fmt::Result {
+        Ok(())
+    }
+
+    fn after_block(&self, _block: super::Block, _f: &mut fmt::Formatter) -> fmt::Result {
+        Ok(())
+    }
+
+    fn before_function_body(&self, _f: &mut fmt::Formatter) -> fmt::Result {
+        Ok(())
+    }
+
+    fn after_function_body(&self, _f: &mut fmt::Formatter) -> fmt::Result {
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+pub struct NOPPrintDecorator();
+impl PrintDecorator for NOPPrintDecorator {}
```

Comment:
as in comment above: hopefully we won't need this if we can make the distinction at the API level (with/without decorator).

Reply:
We still need a `NOPPrintDecorator` when creating the `FunctionBodyDisplay` object in the `display` and `display_verbose` functions in func.rs. FunctionBodyDisplay takes in a generic parameter of type `PrintDecorator` so the typechecker complains if I don't supply it with `NOPPrintDecorator`.

I could instead have the display function return a `dyn PrintDecorator` but then all the callsites for `display` complain about not knowing the size at compile time since that is required by `writeln!` and `log_trace!`.  
```
    pub fn display<'a>(
        &'a self,
        indent: &'a str,
        module: Option<&'a Module>,
    ) -> FunctionBodyDisplay<'_, dyn PrintDecorator> {
        FunctionBodyDisplay::<dyn PrintDecorator> {
            body: self,
            indent,
            verbose: false,
            module,
            decorator: None,
        }
    }
```